lambda/translcore.ml

```diff
@@ -997,16 +891,8 @@ and transl_exp0 ~in_new_scope ~scopes sort e =
                module.  When that changes, some adjustments may be needed
                here. *)
             List.fold_left (fun (body, pos) id ->
-<<<<<<< HEAD
               Llet(Alias, Lambda.layout_module_field, id,
                    Lprim(mod_field pos, [Lvar oid],
-||||||| merged common ancestors
-              Llet(Alias, Pgenval, id,
-                   Lprim(Pfield pos, [Lvar oid],
-=======
-              Llet(Alias, Pgenval, id,
-                   Lprim(Pfield (pos, Pointer, Mutable), [Lvar oid],
->>>>>>> ocaml/5.1
```

Comment:
I'm leaving this comment to reflect the fact we're discussing this internally. I'm surprised that upstream has `Mutable` but that `mod_field` is implemented with `Reads_agree` -- shouldn't it be `Reads_vary`?

Reply:
cc @mshinwell for context on internal conversation

Reply:
I haven't seen the internal conversation, but I assume that it revolves around recursive modules. In which case, I'll point to a previous analysis I did when reviewing a PR on recursive modules: https://github.com/ocaml/ocaml/pull/10205#pullrequestreview-674345642.
My opinion here is that using immutable reads (`Reads_agree`) should be preferred even if it could lead to observable differences:
- For functions, only physical equality can tell the difference (the stub function will behave as the new function if the module has been initialized)
- For lazy values, the old value could leak. This only matters if the lazy value was forced before initialization but the corresponding exception was caught and ignored; otherwise, the old value and the new value behave the same way after initialization.
- For classes, I think the old value could leak, but I haven't done enough tests to know whether it is actually a problem (the language doesn't allow to manipulate class values directly, so it's a bit hard to trigger the complex cases).
- For submodules, no observable difference for the modules themselves, as the modules fields are not patched. The fields of the submodules can be affected though.

Reply:
>     * For classes, I think the old value could leak, but I haven't done enough tests to know whether it is actually a problem (the language doesn't allow to manipulate class values directly, so it's a bit hard to trigger the complex cases).

Update: I just checked, and classes are actually patched in place, so the old value and the new value are the same.

Reply:
OK, so it sounds like it's fine to do `Reads_agree` here (which is what the code currently does). I'll resolve this, thanks. (Though I suspect others may occasionally run into this confusion as long as the Jane Street branch disagrees with `ocaml/ocaml` &mdash; should something be done to make them agree?)